packages/react/src/hooks/dashboard/useDispatchIntent.ts

```diff
@@ -0,0 +1,166 @@
+import {SDK_CHANNEL_NAME, SDK_NODE_NAME} from '@sanity/message-protocol'
+import {type FrameMessage} from '@sanity/sdk'
+import {useCallback} from 'react'
+
+import {useWindowConnection} from '../comlink/useWindowConnection'
+import {type DocumentHandleWithSource} from './types'
+import {getResourceIdFromDocumentHandle} from './utils/getResourceIdFromDocumentHandle'
+
+/**
+ * Message type for sending intents to the dashboard
+ */
+interface IntentMessage {
+  type: 'dashboard/v1/events/intents/dispatch-intent'
+  data: {
+    action?: 'edit'
+    intentId?: string
+    document: {
+      id: string
+      type: string
+    }
+    resource?: {
+      id: string
+      type?: 'mediaLibrary'
+    }
+    params?: Record<string, string>
+  }
+}
+
+/**
+ * Return type for the useDispatchIntent hook
+ * @beta
+ */
+interface DispatchIntent {
+  dispatchIntent: () => void
+}
+
+/**
+ * Parameters for the useDispatchIntent hook
+ * @beta
+ */
+type UseDispatchIntentParams =
+  | {
+      action: 'edit'
+      intentId?: never
+      documentHandle: DocumentHandleWithSource
+      params?: Record<string, string>
+    }
+  | {
+      action?: never
+      intentId: string
+      documentHandle: DocumentHandleWithSource
+      params?: Record<string, string>
+    }
+
+/**
+ * @beta
+ *
+ * A hook for dispatching intent messages to the Dashboard with a document handle.
+ * This allows applications to signal intent to navigate to applications that can perform specific actions on a document.
+ *
+ * @param params - Object containing:
+ *   - `action` - Action to perform (currently only 'edit' is supported). Will prompt a picker if multiple handlers are available.
+ *   - `intentId` - Specific ID of the intent to dispatch. Either `action` or `intentId` is required.
+ *   - `documentHandle` - The document handle containing document ID, type, and either:
```

Comment:
I could be wrong, but looking at the `DocumentHandleWithSource` type, it seems like `projectId` and `dataset` aren't optional; but considering the description below, it seems like perhaps they should be if `source` is provided?